### PR TITLE
feat(exec): add --env to keep credentials off the command line

### LINKED ISSUE #263: Add --env support to exec and warn against command-line credentials in exec/websh help
## Background

AI agents such as Claude Code connect to remote databases through alpacon and often pass credentials directly on the command line, e.g. `mysql -p<password>`. These commands are recorded verbatim in the audit log, which can potentially expose customer secrets.

Passing credentials through `--env` instead keeps the value off the command line. The safest form is to set the secret as a shell environment variable and pass only its key name (`--env="DB_PASS"`), so neither the client-side command line nor the audit log ever contains the actual value—only the reference. Passing `--env="DB_PASS=hunter2"` still writes the value on the local command line, so it should be avoided.

Testing shows `--env` is supported by websh only—exec does not support it at all.

## Observed behavior

websh supports both `--env` forms:

```
# Preferred: the value stays in the shell env, only the key is on the command line
export DB_PASS=hunter2
alpacon websh --env="DB_PASS" web-editor 'echo "pass=$DB_PASS"'
pass=hunter2

# Discouraged: the value is written on the command line
alpacon websh --env="DB_PASS=hunter2" web-editor 'echo "pass=$DB_PASS"'
pass=hunter2
```

exec supports neither, and fails at parse time:

```
alpacon exec --env="DB_PASS=hunter2" web-editor 'echo "pass=$DB_PASS"'
Error: unknown flag: --env=DB_PASS=hunter2
```

Since AI agents primarily use the exec path, there is currently no way to move credentials off the command line for them.

## Changes

1. Add `--env` support to exec
   - websh and exec share the same command-execution path (`RunCommandWithRetry`), which already takes an `env map[string]string` argument. Today `cmd/exec/exec.go:117` always passes an empty map.
   - Reuse websh's `--env` parser (`extractEnvValue`, `cmd/websh/websh.go:337`) to add an `--env` case to `ParseRemoteExecArgs` (`cmd/exec/parse.go`).
   - Support both `--env="KEY=VALUE"` and `--env="KEY"` (use the current shell's value) forms, matching websh.

2. Recommend the shell-env form in exec/websh help text
   - Both forms already exist in websh (`websh.go:160-161`) but are listed as equals. Update the Long/Example text for exec and websh to recommend setting the secret as a shell environment variable and passing only the key (`--env="KEY"`), and to warn that credentials should never be written directly on the command line (including the `--env="KEY=VALUE"` form).

## Testing

- Add unit tests for `--env` parsing in `ParseRemoteExecArgs` (`--env=KEY=VALUE`, `--env=KEY`, malformed input).
- Verify that a credential passed via `--env` to exec is delivered as an environment variable rather than on the command line.

### PR DESCRIPTION
**Pull request by jisung-02**
## Background

AI agents such as Claude Code connect to remote databases through `alpacon exec` and often pass credentials directly on the command line (e.g. `mysql -p<password>`). Those commands are recorded verbatim in the audit log, which can expose customer secrets. Passing the value through `--env` instead keeps it off the alpacon command line. The safest form is to set the secret as a shell environment variable and pass only its key name (`--env="DB_PASS"`), so neither the alpacon command line nor the audit log ever contains the actual value, only a reference.

Previously `--env` was supported by websh only; exec failed at parse time with `unknown flag`. Since AI agents primarily use the exec path, they had no way to move credentials off the command line.

## Changes

Add a shared `ParseEnvArg` helper to the exec package that returns an error string instead of calling `os.Exit`, add an `--env` case to `ParseRemoteExecArgs`, and carry the values through `RemoteExecArgs.Env`. `cmd/exec/exec.go` passes `parsed.Env` into `SubmitCommand`/`RunExecWithApprovalWait` instead of an empty map.

The supported forms match websh. `--env="KEY"` reads the value from the current shell (only the key name is on the command line); `--env="KEY=VALUE"` sets a literal value.

`--env` detection matches only the `--env` and `--env=` forms, so an unrelated flag like `--env-file=...` is rejected as an unknown flag by exec (and, in websh, falls through to the server-name slot, which has no unknown-flag rejection) rather than being swallowed and reported as a malformed `--env`.

### Shared parsing and execution across exec and websh

The `--env` parser is unified: websh's `--env` case delegates to the same `ParseEnvArg`, so both commands accept identical forms and produce identical malformed-input errors. `ParseEnvArg` is a pure parser—it returns the error string, and both `ParseRemoteExecArgs` (exec) and `ParseWebshArgs` (websh) surface it the same way, through their normal return path rather than exiting inside the parser.

The execution path is also shared. The post-parse body of exec's command is extracted into `RunRemoteExec`, and websh command mode (`websh SERVER COMMAND`) now maps its parsed args onto `RemoteExecArgs` and calls `RunRemoteExec`. As a result, running a command through websh behaves like `alpacon exec`: same output, exit-code propagation, sudo-approval handling (including the pending-approval exit code 4 and the presence step-up), and `--output json` buffering. websh does not parse exec-only flags such as `--detach`/`--wait`, so its `Wait` is always false and the blocking approval wait never runs; the interactive `websh SERVER` path (no command) is unchanged and keeps its own sudo MFA listener.

The exec and websh help text recommends the shell-env form (`--env="KEY"`), warns that the `--env="KEY=VALUE"` form writes the value on the command line and into the audit log, and scopes the secrecy claim to the alpacon command line (the local `export` that sets the value is still on the shell command line).

## Safety

The whole point of the feature is to keep the secret off the alpacon command line and out of the audit log, so I verified it is never re-exposed. The secret value flows only through the `Env` map; it never enters the `ShellJoin`/`Command` string, stdout, or error/warning messages. The unset-variable warning and the malformed-input error print only the key name, never the value.

`reRunHint` reconstructs the invocation for the pending-approval message and now includes the `--env` keys so a re-run stays faithful—but it emits `--env=KEY` only, never `--env=KEY=VALUE`. The re-run re-reads each value from the shell, so no value lands in the hint on stderr or in the logs.

The help examples use patterns that do not re-expose the value. `PGPASSWORD` with `psql` reads the value from the environment rather than argv, so it lands neither in the remote process's argv (visible via `ps`) nor in terminal output / the session recording.

One behavior change: in websh, malformed `--env` input (an empty key, e.g. `--env=`) now fails the command instead of the previous warn-and-skip. Failing loudly on a malformed credential flag is safer than silently dropping it, and it is consistent with exec.

## Testing

Table-driven coverage for `--env` parsing in both `ParseRemoteExecArgs` and `ParseWebshArgs`: both forms (`KEY=VALUE`, shell-sourced `KEY`), quote unwrapping, unset-variable skip, multiple flags accumulating, empty value, malformed input, and `--env` after the server name being treated as a command arg. It also asserts the shell-sourced value lands in the env map and never appears in the parsed command string. Dedicated regression tests cover the secret-not-echoed-on-malformed-input path (`TestParseRemoteExecArgs_EnvErrorHidesSecret`, `TestParseWebshArgs_EnvErrorHidesSecret`), the exact `--env` matching (`TestParseRemoteExecArgs_EnvPrefixIsExact`, `TestParseWebshArgs_EnvPrefixIsExact`), and the keys-only rerun hint (`TestReRunHint`).

Manual end-to-end run against a live workspace (server `web-editor`):

- Parse-time (no server call): `--env-file=...` rejected as unknown flag in exec; malformed `--env='=hunter2'` returns the format error with the value never echoed, in both exec and websh.
- `--env="PGPASSWORD"` injects the shell value into the remote process (`echo "$PGPASSWORD"` prints it) with only the key on the local command line; `--env=A=1 --env=B=2` accumulates; an unset key warns and is skipped.
- websh command mode matches exec: `--env` injection works, and `exit 7` propagates as exit code 7 through both `exec` and `websh`; `--output json` produces identical output.

`go build ./...`, `go test -race ./cmd/exec/... ./cmd/websh/...`, and `golangci-lint run ./...` all pass.

## Checklist

- [x] `go build ./...` passes
- [x] `go test -race ./cmd/exec/... ./cmd/websh/...` passes
- [x] `golangci-lint run ./...` reports 0 issues
- [x] exec/websh help text updated
- [x] Manual e2e verified against a live workspace

Closes #263

### CODE REVIEW

File: cmd/exec/parse.go [hunk 1]
```diff
@@ -135,11 +143,38 @@ func ParseRemoteExecArgs(args []string) RemoteExecArgs {
...
+func ParseEnvArg(arg string, env map[string]string) string {
+	body, ok := strings.CutPrefix(arg, "--env=")
+	var key, value string
+	var hasValue bool
+	if ok {
+		key, value, hasValue = strings.Cut(strings.Trim(body, "\""), "=")
+	}
+	if !ok || key == "" {
+		return fmt.Sprintf("invalid --env argument %q: use --env=KEY or --env=KEY=VALUE", arg)
+	}
```
[the remark this answers is not in the document]
Comment (jisung-02) [reply cmd/exec/parse.go:163]: Good catch—fixed in 2305068. The malformed-input branch no longer formats the raw argument; it returns only the expected format (`invalid --env argument: use --env=KEY or --env=KEY=VALUE`), so a botched secret-bearing token like `--env="=hunter2"` never reaches stderr or logs. Added `TestParseRemoteExecArgs_EnvErrorHidesSecret` to lock it in, and websh inherits the fix through the shared `ParseEnvArg` helper.

File: cmd/exec/parse.go [hunk 2]
```diff
@@ -98,6 +101,10 @@ func ParseRemoteExecArgs(args []string) RemoteExecArgs {
 			if outputFormat == "" {
 				return RemoteExecArgs{Err: "--output requires a value (table|json)"}
 			}
+		case strings.HasPrefix(arg, "--env"):
+			if errMsg := ParseEnvArg(arg, env); errMsg != "" {
+				return RemoteExecArgs{Err: errMsg}
+			}
```
[the remark this answers is not in the document]
Comment (jisung-02) [reply cmd/exec/parse.go]: Fixed in 6ed527c. `--env` now matches only the `--env` and `--env=` forms (`arg == "--env" || strings.HasPrefix(arg, "--env=")`), mirroring `--output` and `--work-session`. `--env-file=...` now falls through to the unknown-flag branch in exec (and to the server-name slot in websh, which has no unknown-flag rejection). Regression tests `TestParseRemoteExecArgs_EnvPrefixIsExact` and `TestParseWebshArgs_EnvPrefixIsExact` lock it in.

File: cmd/websh/websh.go [hunk 3]
```diff
@@ -119,8 +120,12 @@ Shell metacharacters (;, |, &, $) pass through unquoted to the remote shell.
 To send a literal metacharacter, wrap the argument in quotes:
   alpacon websh server 'echo hello;world'
 
+Executing a command (SERVER followed by COMMAND) behaves exactly like 'alpacon exec'.
+
```
[the remark this answers is not in the document]
Comment (jisung-02) [reply cmd/websh/websh.go]: Fixed in ae35c8f. The help text no longer claims full exec equivalence. It now reads that command execution runs through the same executor as `alpacon exec`—identical output, exit codes, and sudo-approval handling—and explicitly states that exec-only flags such as `--detach`/`--wait` are not accepted. The delegation comment was corrected for the same reason (`Wait` is always false in websh, so the blocking approval wait never runs).

File: cmd/exec/exec.go [hunk 4]
```diff
@@ -97,88 +110,94 @@ Requires an active WorkSession when using Browser login (Auth0); Token auth (API
...
+	// A sudo command pending human approval (SUDO_APPROVAL_REQUIRED) that we did
+	// not --wait on emits a machine-readable pending signal and exits before the
+	// normal result handling treats the denial as a plain failure.
+	if HandlePendingApproval(err, reRunHint(parsed)) {
+		return
+	}
+	if buf != nil {
+		_, _ = os.Stdout.Write(buf.Bytes())
+	}
+	HandleCommandResult(err)
```
[the remark this answers is not in the document]
Comment (jisung-02) [reply cmd/exec/exec.go:201]: Fixed in 74cb006. `reRunHint` now emits `--env=KEY` for each key in `parsed.Env`, sorted for stable output, placed before the server (matching exec's flag layout). Keys only, never values: the rerun re-reads each value from the shell, so a secret never lands in the hint on stderr or in the logs. `TestReRunHint` gained an `emits env keys sorted, never values` case asserting both the exact reconstruction and the absence of the secret values.

File: cmd/exec/exec.go [hunk 5]
```diff
@@ -38,6 +39,13 @@ To send a literal metacharacter, wrap the argument in quotes:
 Flags:
   -u, --username [USER_NAME]    Specify the username for command execution.
   -g, --groupname [GROUP_NAME]  Specify the group name for command execution.
+  --env="KEY"                   Pass an environment variable to the remote command,
+                                reading its value from the current shell. This keeps
+                                the value off the command line and out of the audit
+                                log—use it for secrets such as passwords or tokens.
```
[the remark this answers is not in the document]
Comment (jisung-02) [reply cmd/exec/exec.go]: Fixed in 481e9ce. Reworded to "off the alpacon command line and out of the audit log", so the claim no longer implies the value can never appear on any command line (the local `export` that sets it does).

File: cmd/exec/exec.go [hunk 6]
```diff
@@ -66,6 +74,12 @@ Requires an active WorkSession when using Browser login (Auth0); Token auth (API
   alpacon exec -u root prod-docker systemctl status nginx
   alpacon exec -g docker user@server docker images
 
+  # Pass a secret via the shell env; the value stays off every command line.
```
[the remark this answers is not in the document]
Comment (jisung-02) [reply cmd/exec/exec.go]: Fixed in 481e9ce. The example comment now reads "stays off the alpacon command line and out of the audit log" instead of "off every command line", which was inaccurate given the `export PGPASSWORD=...` shown just below it.

File: cmd/websh/websh.go [hunk 7]
```diff
@@ -136,8 +143,10 @@ Requires an active WorkSession when using Browser login (Auth0); Token auth (API
   alpacon websh my-server "ls -la /var/log"
   alpacon websh root@my-server "systemctl status nginx"
 
-  # Set environment variables
-  alpacon websh --env="KEY1=VALUE1" --env="KEY2=VALUE2" my-server "echo $KEY1"
+  # Pass a secret via the shell env; the value stays off every command line and
+  # out of the session recording. psql reads PGPASSWORD directly from the env.
```
[the remark this answers is not in the document]
Comment (jisung-02) [reply cmd/websh/websh.go]: Fixed in 481e9ce. Reworded to "off the alpacon command line and out of the session recording"; the `export PGPASSWORD=...` line does put the value on the local shell command line, so the previous "off every command line" overstated it.

File: cmd/websh/websh.go [hunk 8]
```diff
@@ -157,8 +166,14 @@ Requires an active WorkSession when using Browser login (Auth0); Token auth (API
 Flags:
   -u, --username [USER_NAME]         Specify the username for command execution.
   -g, --groupname [GROUP_NAME]       Specify the group name for command execution.
-  --env="KEY=VALUE"                  Set environment variable 'KEY' to 'VALUE'.
-  --env="KEY"                        Use the current shell's value for 'KEY'.
+  --env="KEY"                        Pass an environment variable, reading its value
+                                     from the current shell. This keeps the value off
+                                     the command line and out of the audit log—use it
+                                     for secrets such as passwords or tokens.
```
[the remark this answers is not in the document]
Comment (jisung-02) [reply cmd/websh/websh.go]: Fixed in 481e9ce. The websh `--env="KEY"` description now says "off the alpacon command line and out of the audit log", matching the exec wording and avoiding the implication that it stays off every command line including the local `export`.

File: cmd/exec/parse.go [hunk 9]
```diff
@@ -135,11 +142,38 @@ func ParseRemoteExecArgs(args []string) RemoteExecArgs {
...
 }
 
+// ParseEnvArg parses a --env token into env. A bare KEY reads the shell value,
+// warning and skipping if unset; malformed input returns an error message.
+func ParseEnvArg(arg string, env map[string]string) string {
+	body, ok := strings.CutPrefix(arg, "--env=")
+	var key, value string
+	var hasValue bool
+	if ok {
+		key, value, hasValue = strings.Cut(strings.Trim(body, "\""), "=")
```
Comment (eunyoung14) [inline cmd/exec/parse.go]: **Suggestion:** `strings.Trim(body, "\"")` strips ALL leading and trailing double quotes unconditionally (not one matched pair), and it runs before the `=` split. Verified corruption cases:

- `--env=PW=trailing"` → value `trailing` (silent data loss for a value legitimately ending in `"`)
- `--env=KEY="VAL"` (quotes preserved into argv by a script) → value `"VAL` (leading kept, trailing stripped)
- `--env="KEY"=VALUE` → key `KEY"`

The common secret path (`--env=KEY` shell-read) is unaffected, and values with internal quotes or `=` parse correctly, so blast radius is limited to the discouraged literal form — but it is silent corruption with no diagnostic. Suggest stripping at most one matched leading+trailing quote pair (or dropping the Trim entirely — a real shell already removes the wrapping quotes before argv). Also note the existing "quoted --env value is unwrapped" test only covers the symmetric-wrap happy path; none of the cases above is covered.

File: cmd/websh/websh.go [hunk 10]
```diff
@@ -223,58 +254,51 @@ Note: All flags must be placed before the server name.
...
 		}
 
-		if len(commandArgs) > 0 {
-			command := execCmd.ShellJoin(commandArgs)
-			err := execCmd.RunCommandWithRetry(alpaconClient, serverName, command, username, groupname, env, workSessionID, os.Stdout)
-			utils.HandleWorkSessionError(err, "command", serverName, authMethod, workSessionID)
-			execCmd.HandleCommandResult(err)
-		} else {
-			session, err := websh.CreateWebshSession(alpaconClient, serverName, username, groupname, share, readOnly, workSessionID)
+		session, err := websh.CreateWebshSession(alpaconClient, serverName, username, groupname, share, readOnly, workSessionID)
```
Comment (eunyoung14) [inline cmd/websh/websh.go:263]: **Suggestion:** interactive websh still parses `--env` but drops it — `CreateWebshSession` takes no env, so `alpacon websh --env=PGPASSWORD my-server` (no command) opens a terminal without the variable and without any warning. This is pre-existing, but this PR's help text now prominently frames `--env` as the secrets channel without scoping it to command mode, so the silent no-op became much more discoverable. Suggest a warning when `len(env) > 0 && len(commandArgs) == 0` (e.g. "--env has no effect on an interactive websh session; it applies only when running a command"), or scoping the help text.

File: cmd/exec/exec.go [hunk 11]
```diff
@@ -196,6 +216,16 @@ func reRunHint(parsed RemoteExecArgs) string {
...
 	}
+	// Emit env keys only (never values): the rerun re-reads each value from the
+	// shell, so the secret stays off this hint on stderr and out of the logs.
+	keys := make([]string, 0, len(parsed.Env))
+	for k := range parsed.Env {
+		keys = append(keys, k)
+	}
+	sort.Strings(keys)
+	for _, k := range keys {
+		parts = append(parts, "--env="+k)
```
Comment (eunyoung14) [inline cmd/exec/exec.go:231]: **Note:** for an invocation that used the literal form (`--env=KEY=VALUE`), this keys-only hint is not replay-faithful: the re-run re-reads `KEY` from the shell, where it is typically unset → warn-and-skip → the replayed command submits without that env key. A human sees the stderr warning; an AI agent replaying the JSON `next_actions[0].command` verbatim (this PR's stated audience) may not. The keys-only choice is the right security tradeoff — just consider one caveat line in the pending-approval message ("--env values are re-read from your shell; export them before re-running") so the failure mode is diagnosed at the right moment.

File: README.md [hunk 12]
```diff
@@ -291,7 +291,7 @@ What each refusal code means and what to do next:
...
+| `4`  | Pending human approval—the action is awaiting an out-of-band approve/reject in the Alpacon console (web/Slack), not refused. For `exec`, re-run the command after approval (or pass `--wait` on the original command to block); for `work-session create` the session already exists—after approval attach it with `alpacon work-session use <id>` (or pass `--wait` on the original create to block). Under `--output json`, a `{"status":"pending_approval", ...}` object is emitted. Returned by `exec` (and `websh` when running a command) on a `SUDO_APPROVAL_REQUIRED` sudo denial and by `work-session create` when the session lands pending |
```
Comment (eunyoung14) [inline README.md]: **Nit:** the row now correctly says exit 4 is returned by `websh` when running a command too, but the guidance "or pass `--wait` on the original command to block" doesn't apply to websh (it has no `--wait` — the in-CLI websh help gets this right by pointing at `alpacon exec --wait`). Consider mirroring that wording here.

**Review by eunyoung14 — COMMENTED** [review]
## Review: --env for exec (cross-repo verified)

### Verification

- Local: `go build ./...` and `go test -race ./cmd/exec/... ./cmd/websh/...` pass on the PR head.
- **Env flow to the server verified end-to-end**: `parsed.Env` (always non-nil) reaches `CommandRequest.Env` on both the detach and streaming paths; every `SubmitCommand` caller threads env; empty map serializes identically to the old `make(map[string]string)`. The detach path previously dropped env entirely — now correctly threaded.
- **Server/agent chain**: the server stores the map on `Command.env` and alpamon injects it into the remote process; audit masking of these values is covered by merged alpacon-server #2450 and extended to live chunks by #2489. The secrecy design here composes correctly with both.
- **websh delegation regressions swept**: single `HandleWorkSessionError` call, same `ResolveOrExit`, identical connection-failure/output-format messages; the behaviors command mode gains (presence step-up, exit code 4 + JSON envelope, JSON buffering) are all documented. Clean.

### Notes (no inline)

- `--env KEY` (space-separated) is a hard error while every sibling flag (`-u`, `--output`, `--work-session`) accepts the space form via extractValue. Defensible (keeps the token adjacent and unambiguous), but an operator will trip on it — consider a one-line mention in the help text.
- A websh-initiated pending approval emits an `alpacon exec ...` rerun hint. Given the help now documents command mode as an exec alias (and websh has no `--wait`), this reads as by-design — fine as is.

**Verdict: COMMENT** — no blockers; the two inline suggestions (quote trimming, interactive --env no-op) are the ones worth acting on before merge.

File: cmd/exec/parse.go (hunk 9 above)
Comment (jisung-02) [reply cmd/exec/parse.go]: Fixed in 7ece05f. Replaced the unconditional `strings.Trim(body, "\"")` with `trimMatchedQuotes`, which strips at most one balanced leading+trailing pair and leaves unbalanced quotes intact — so `--env=PW=trailing"` keeps its value and `--env=KEY="VAL"` passes through verbatim instead of being half-stripped. Kept the existing symmetric-wrap unwrap, and added the trailing-quote and inner-quoted cases to `parse_test.go` since they were previously uncovered.

File: cmd/websh/websh.go (hunk 10 above)
Comment (jisung-02) [reply cmd/websh/websh.go:263]: Fixed in 1e8b37a. The interactive path now warns when `--env` keys are present (`--env has no effect on an interactive websh session; it applies only when running a command`) instead of silently dropping them, so the no-op is diagnosed at the moment it happens.

File: cmd/exec/exec.go (hunk 11 above)
Comment (jisung-02) [reply cmd/exec/exec.go:231]: Addressed in 757b241. `reRunHint` now returns a `utils.NextAction`: the `Command` stays a pure, executable string (so a machine consumer can replay it directly), and when `--env` keys are present the caveat "--env values are re-read from your shell; export them before re-running" rides in `Description`. That surfaces in both the human stderr line (`command  # description`) and the JSON `next_actions[0].description`, so an agent replaying the command still gets the caveat.

File: README.md (hunk 12 above)
Comment (jisung-02) [reply README.md]: Fixed in 830b74b. The exit-code 4 row now adds "`websh` command mode has no `--wait`, so re-run via `alpacon exec --wait` to block", mirroring the in-CLI websh help.

**Review by eunyoung14 — APPROVED** [review]
## Re-review: all points addressed — approving

Verified each fix on the new head:

- **Quote trimming** (7ece05f): `trimMatchedQuotes` strips exactly one balanced wrapping pair and leaves unbalanced quotes intact — `--env=PW=trailing"` no longer loses its closing quote, the documented symmetric unwrap is preserved, and the previously-uncovered cases now have tests.
- **Interactive websh --env** (1e8b37a): the silent no-op now warns at the moment it happens. 
- **Rerun-hint caveat** (757b241): making `reRunHint` return a `NextAction` so the caveat rides in `Description` — surfacing in both the human stderr line and the JSON `next_actions[0].description` while keeping `Command` a pure runnable string — covers the machine consumer better than what I suggested. Nice.
- **README exit-code 4 row** (830b74b): now mirrors the in-CLI websh guidance.

Local re-verification: `go build ./...` and `go test -race ./cmd/exec/... ./cmd/websh/...` pass.

**Verdict: APPROVE**
